crates/chaintsn/src/transition.rs

```diff
@@ -86,51 +88,69 @@ fn compute_init_slot_rng(state: &StateCache) -> SlotRng {
 ///
-/// Returns if there was an update processed.
+/// Returns true if there epoch needs to be updated.
 fn process_l1_view_update(
     state: &mut StateCache,
     l1seg: &L1Segment,
     params: &RollupParams,
 ) -> Result<bool, TsnError> {
     let l1v = state.state().l1_view();
 
-    // Accept new blocks.
-    if !l1seg.new_manifests().is_empty() {
-        let cur_safe_height = l1v.safe_height();
+    if l1seg.new_manifests().is_empty() {
+        return Ok(false);
+    }
 
-        // Validate the new blocks actually extend the tip.  This is what we have to tweak to make
-        // more complicated to check the PoW.
-        let new_tip_height = cur_safe_height + l1seg.new_manifests().len() as u64;
-        // FIXME: This check is just redundant.
-        if new_tip_height <= l1v.safe_height() {
-            return Err(TsnError::L1SegNotExtend);
-        }
+    let cur_safe_height = l1v.safe_height();
 
-        // Go through each manifest and process it.
-        for (off, b) in l1seg.new_manifests().iter().enumerate() {
-            // PoW checks are done when we try to update the HeaderVerificationState
-            let header: Header =
-                consensus::deserialize(b.header()).expect("invalid bitcoin header");
-            state.update_header_vs(&header, &Params::new(params.network))?;
+    // Validate the new blocks actually extend the tip.  This is what we have to tweak to make
+    // more complicated to check the PoW.
+    let new_tip_height = cur_safe_height + l1seg.new_manifests().len() as u64;
+    // FIXME: This check is just redundant.
+    if new_tip_height <= l1v.safe_height() {
+        return Err(TsnError::L1SegNotExtend);
+    }
 
-            let height = cur_safe_height + off as u64 + 1;
-            process_l1_block(state, b)?;
-            state.update_safe_block(height, b.record().clone());
-        }
+    let prev_finalized_epoch = *state.state().finalized_epoch();
+
+    // Go through each manifest and process it.
+    for (off, b) in l1seg.new_manifests().iter().enumerate() {
+        // PoW checks are done when we try to update the HeaderVerificationState
+        let header: Header = consensus::deserialize(b.header()).expect("invalid bitcoin header");
+        state.update_header_vs(&header, &Params::new(params.network))?;
+
+        let height = cur_safe_height + off as u64 + 1;
+        process_l1_block(state, b, params)?;
+        state.update_safe_block(height, b.record().clone());
+    }
+
+    // If prev_finalized_epoch is null, i.e. this is the genesis batch, it is safe to update the
+    // epoch
+    if prev_finalized_epoch.is_null() {
+        return Ok(true);
+    }
+
+    // For all other non-genesis batch, we need to check that the new finalized epoch has been
+    // updated when processing L1Checkpoint
+    let new_finalized_epoch = state.state().finalized_epoch();
 
+    if new_finalized_epoch.epoch() > prev_finalized_epoch.epoch() {
```

Comment:
the only concern with this check is that, this implies we can have l1 segment without checkpoint, which is not true in our case because we only create a l1 segment if we see a valid previous checkpoint on L1. In practice, this probably won't cause issues, but seems like this kind of violates the theoretical invariant.